### ISSUE TITLE
Every visit to the HomeFragment triggers a full account sync

### ISSUE THREAD

**Issue by jonalmeida** [mozilla-mobile/fenix#26198]
In https://github.com/mozilla-mobile/fenix/pull/26028, we've added the method `collectAccountUpdates`. This method calls `accountManager.syncNow` and `accountManager.withConstellation { refreshDevices() }` and is called for every `onStart`/`onPause` of the fragment because it's a `LifecycleAwareFeature`.

This is problem because even opening the tabs tray and dismissing it (with the home fragment in the background triggers a sync which leads to multiple GET and POST requests:
<img width="466" alt="Screen Shot 2022-07-26 at 10 05 38 PM" src="https://user-images.githubusercontent.com/1370580/181144733-759f3349-1c2b-459c-85c7-01da433cde33.png">

I would have expected that we allow our regular sync polling to update the SyncStore and this feature would observe the state and display the most recent synced tab that we already have. 

cc: @MatthewTighe

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21123)

**Comment by jonalmeida** [comment]
If this issue is a duplicate, please close it. If not, could we bump this up in sprint priority? We would be putting a lot more load on Sync servers and wouldn't be great for limited bandwidth users.

**Comment by MatthewTighe** [comment]
Thanks for bringing this to my attention! Was this caught by a general monitoring strategy, or is this something you check personally with some regularity? Just curious if there are additional tools or anything for me to keep my eye on

I believe that the underlying behavior (the recent synced tab is refreshed on every visit to the home screen) is actually what's desired for the feature. If that's a bigger deal for the Sync servers than we'd thought, I should run it by product again. Has the Sync team warned about this kind of usage in the past?

> In https://github.com/mozilla-mobile/fenix/pull/26028, we've added the method collectAccountUpdates. This method calls accountManager.syncNow and accountManager.withConstellation { refreshDevices() } and is called for every onStart/onPause of the fragment because it's a LifecycleAwareFeature.

This shouldn't actually be different than the previous implementation. Before, the `RecentSyncedTabFeature` had implemented `SyncedTabsFeature` which calls `DefaultController.syncAccount` in the feature's lifecycle-aware`start` method. The block in question in the OP is copied from `syncAccount`.

**Comment by Mugurell** [comment]
The issue about downloading new data everytime the user goes to homescreen was discussed in today's planning meeting.
Looking over the current telemetry we see that we have:
- 23 million `tabs_tray.new_private_tab_tapped` events showing the homescreen - https://mozilla.cloud.looker.com/explore/fenix/event_counts?qid=eVfRSqGWSujpYGaohvbJCz&toggle=vis
- 107 million `tabs_tray.new_tab_tapped events` showing the homescreen - https://mozilla.cloud.looker.com/explore/fenix/event_counts?qid=XFVRcgoptR4lg3q4FrAIcr&toggle=vis
- one billion `home_screen_home_screen_displayed` total events - https://mozilla.cloud.looker.com/explore/fenix/event_counts?qid=1ibIlMB3KoVSqxp1a7wVGR&toggle=vis

Don't know about the exact specifics but some general approaches in easing the server pressure would be:
- implementing a cooldown period locally - quick and easy
- add support for delta/incremental syncs (probably the best compromise)
- as a better scaling solution for the future, to really avoid making new requests to then just disregard downloaded data which is identical to which we have locally we could switch to a push based arch with using a socket through which the server sends delta/incremental syncs. (definitely a big change requiring important work)

**Comment by MatthewTighe** [comment]
I'm surprised that the difference between `tabs_tray.new_tab_tapped` events and `home_screen_home_screen_displayed` is so large. Generally speaking, it's seemed that we've treated the home screen as a "users don't see this very often" area of the app since I've been here though I know the intent is to change that. 

It looks like just clicking the Awesomebar now opens the home screen as well. Could that be contributing? Looking for the metrics now.

It might make the most sense to just differentiate between user intent. If a user is clicking the home button, I think it's fair to expect a sync to trigger so that they get the latest data. A new tab click is more questionable, and I would personally not expect tapping the Awesomebar to trigger it. It may make the most sense to only trigger the sync in specific cases of where the home screen is being opened from.

**Comment by jonalmeida** [comment]
> Thanks for bringing this to my attention! Was this caught by a general monitoring strategy, or is this something you check personally with some regularity? Just curious if there are additional tools or anything for me to keep my eye on

I was investigating another issue and noticed this peculiarity while using `about:debugging` to inspect the Main process with the Network tab open (which is where my screenshot was from in the opening comment).

> Has the Sync team warned about this kind of usage in the past?

Yes, but this can sometimes be a delayed response because a trend graph would show the usages spike only _after_ this change makes it's way into the release app and then the investigation from the Sync team that ends with us investigating it too. If we are considerate with our resources, we can avoid DDOS-ing our own servers. 😄 

> implementing a cooldown period locally - quick and easy

@Mugurell @MatthewTighe This exists (for some part)! In the `syncNow` API, we have a `debounce` param. Setting that will, as described, reduce multiple requests to just one in a time interval. What we could possibly do, is also change that default value to be true. This might avoid multiple sync requests since as Matt hinted above, we do make other sync requests prior to this change and also when the tabs tray is opened.

> I'm surprised that the difference between tabs_tray.new_tab_tapped events and home_screen_home_screen_displayed is so large. 

You can enter (or re-enter) the homescreen from other places as well e.g. visit settings then navigate back, or switch between Fenix and other apps.

> Generally speaking, it's seemed that we've treated the home screen as a "users don't see this very often" area of the app since I've been here though I know the intent is to change that.

I don't think "users don't see this very often" was ever the intention otherwise we wouldn't be trying to make it more useful. 🙃

**Comment by MatthewTighe** [comment]
> You can enter (or re-enter) the homescreen from other places as well e.g. visit settings then navigate back, or switch between Fenix and other apps.

I think all of these examples could be ones that would have an expectation of refreshing the latest data, but I'd defer to product. Generally, I think if a user is intentionally moving to the home screen it may be fair to assume they want this feature to show them their most recent synced tab.

The magnitude of requests might be a little surprising to the Sync team, but this "task continuity" effort has been a large, planned, crossplatform one. Desktop, for example, is planning to initiate syncs every 5 seconds IIRC, in order to ensure we have always have access to the latest data. If I had to guess, the volume there will be a lot higher than whatever we see on mobile. Debouncing to that 5 second value could be an easy answer on mobile, for sure. We should also remember that only a fraction of those visits will be from users with Sync enabled.

> > Generally speaking, it's seemed that we've treated the home screen as a "users don't see this very often" area of the app since I've been here though I know the intent is to change that.

> I don't think "users don't see this very often" was ever the intention otherwise we wouldn't be trying to make it more useful. 🙃

Apologies if the tone of that read derisively of anything, not my intent. I'm fully onboard with making the home screen more useful, was just trying to communicate why I was surprised by the numbers @Mugurell shared. There are more methods of accessing the home screen than I'd realized or considered in the context of these changes.

**Comment by jonalmeida** [comment]
> I think all of these examples could be ones that would have an expectation of refreshing the latest data, but I'd defer to product. Generally, I think if a user is intentionally moving to the home screen it may be fair to assume they want this feature to show them their most recent synced tab.
> 
> The magnitude of requests might be a little surprising to the Sync team, but this "task continuity" effort has been a large, planned, crossplatform one. Desktop, for example, is planning to initiate syncs every 5 seconds IIRC, in order to ensure we have always have access to the latest data. If I had to guess, the volume there will be a lot higher than whatever we see on mobile. Debouncing to that 5 second value could be an easy answer on mobile, for sure. We should also remember that only a fraction of those home screen visits will be from users with Sync enabled.

@MatthewTighe We can provide users with the most recent synced tabs with reasonable sync-intervals. I'd argue that our sync intervals are different from desktops because the platforms are inherently different, so we should be more prudent with our resources.

I would still recommend we use our debouncer in all the places that we can since the 5 seconds is a reasonable fix and doesn't affect the _perceived value_ to the user.

If you feel strongly, we can check with product and then talk to the Sync team to see if the load is manageable.

**Comment by MatthewTighe** [comment]
I think adding the debounce is a good idea, and will be an immediate improvement towards this issue. I think there would still be a couple open concerns, however. Just want to jot them down here as reference, especially since it sounded like @rocketsroger may pick this one up:

- I'm not sure how frequently visits to the home screen occur within 5 seconds of each other. The one use case that's obvious there is opening the tabs tray, if that does indeed start/stop the fragment. @jonalmeida is that the behavior you saw during your initial testing? I would have expected it to just pause/resume, but haven't verified. If that's the case, another potential easy win would be to move the sync behavior to trigger in `onCreate`, though that may have higher impact on cold starts. 
- This would be a longer-term improvement, but it might be worth investigating ways to remove the `refreshDevices` piece of this. Without that call, retrieval of synced tabs fails because it cannot pair the tabs to remote devices. `refreshDevices` is the mechanism by which the `state` of the `DeviceConstellation` is populated. We could potentially combine those such that the device state is cached appropriately but synced tabs retrieval does not error if the device refresh hasn't completed yet.

**Comment by jonalmeida** [comment]
>     * @jonalmeida is that the behavior you saw during your initial testing?

That is correct.

> I would have expected it to just pause/resume, but haven't verified. If that's the case, another potential easy win would be to move the sync behavior to trigger in `onCreate`, though that may have higher impact on cold starts.

If we move this to onCreate, then we reduce the refresh of tabs to only once in the Activity lifecycle. Since we only have one Activity, that would be far less than what you would probably want for this feature.

If you mean the fragments, onCreate, that could be better, but I would still enable the debouncer, I think. I'm not certain, but we may still be overdoing it depending on how the navgraph decides to cache fragments; needs investigation.

**Comment by jonalmeida** [comment]
To summarize above: we're checking with the Sync servers immediately every time we open the home screen or open/close the tabs tray. We want to be more careful about how often we check by saying, don't check again if you just checked less than 5 secs ago. 

@athomasmoz could you confirm that this solution is acceptable for the feature?

**Comment by athomasmoz** [comment]
Confirmed, thanks!

**Comment by mhammond** [comment]
Note that it will not be necessary to perform a full sync - Android has the capability to ask for only "tabs" to be synced, and I believe it already leverages that when showing the new-tab page. Desktop has a highly optimized tabs-only sync that app-services doesn't yet support, but I think that will be fine in the short term, and we'll loop the server team in as this lands to make sure.

**Comment by jonalmeida** [comment]
@mhammond maybe this bug is not as serious as I imagined (or hopefully invalid).

I see that we are making the `syncNow` call requesting only tabs that are synced, but looking at the endpoints that we hit there are more than I imagined (`token.services.mozilla.com/1.0/sync/1.5`, `/collections`, `/keys`, `/global`). Are those expected to happen before the desired `/clients` and `/tabs`? Is that what you are referring to as the current app-services implementation?

I counted three times this group of requests go out during a syncs right now (when signed in):
 1. The new-tab page. <- new
 2. The tabs tray since we have Synced Tabs there (even without navigating to the Synced Tabs view).
 3. The main settings page (probably because there is account information there to show your name/email/avatar).

So either way, adding the debounce flag will definitely improve the the situation.

**Comment by mhammond** [comment]
> @mhammond maybe this bug is not as serious as I imagined (or hopefully invalid).
> 
> I see that we are making the `syncNow` call requesting only tabs that are synced, but looking at the endpoints that we hit there are more than I imagined (`token.services.mozilla.com/1.0/sync/1.5`, `/collections`, `/keys`, `/global`). Are those expected to happen before the desired `/clients` and `/tabs`? Is that what you are referring to as the current app-services implementation?

Yes, that's correct - https://github.com/mozilla/application-services/issues/5074 is on file to perform the same optimizations as desktop does.

> I counted three times this group of requests go out during a syncs right now (when signed in):
> 
>     1. The new-tab page. <- new
> 
>     2. The tabs tray since we have Synced Tabs there (even without navigating to the Synced Tabs view).

Both of those can probably be just tab syncs, and is maybe worth debouncing (having done a tab sync just X seconds ago is probably fine)

>     3. The main settings page (probably because there is account information there to show your name/email/avatar).

It seems odd that this would be just a tabs sync, and in theory, no sync should be necessary to get that info as it comes from fxa-client rather than from sync. OTOH, I doubt it hurts that much.

**Comment by jonalmeida** [comment]
> Both of those can probably be just tab syncs, and is maybe worth debouncing (having done a tab sync just X seconds ago is probably fine)

Matt has already made these are tab syncs, so they just need to be debounced then - thank you!
